## UK House Prices Analysis

Source
https://www.gov.uk/government/statistical-data-sets/uk-house-price-index-data-downloads-december-

Information on dataset
https://www.gov.uk/government/publications/about-the-uk-house-price-index/about-the-uk-house-price-index#data-tables

In [248]:
import numpy as np
import os
import pandas as pd

In [249]:
filepath = "../data/UK-HPI-full-file-2025-12.csv"

#### Create a dataframe from csv

In [251]:
data = pd.read_csv(filepath)

#### Examine data

In [253]:
print(f"Data shape:{data.shape}")
data.head()

Data shape:(149085, 54)


,Date,RegionName,AreaCode,AveragePrice,Index,IndexSA,1m%Change,12m%Change,AveragePriceSA,SalesVolume,...,NewPrice,NewIndex,New1m%Change,New12m%Change,NewSalesVolume,OldPrice,OldIndex,Old1m%Change,Old12m%Change,OldSalesVolume
0,01/01/2004,Aberdeenshire,S12000034,84638,41.1,NaN,NaN,NaN,NaN,388.0,...,112843.0,40.7,NaN,NaN,103.0,81273.0,41.0,NaN,NaN,285.0
1,01/02/2004,Aberdeenshire,S12000034,84623,41.1,NaN,0.0,NaN,NaN,326.0,...,113061.0,40.8,0.2,NaN,107.0,81194.0,40.9,-0.1,NaN,219.0
2,01/03/2004,Aberdeenshire,S12000034,86536,42.1,NaN,2.3,NaN,NaN,453.0,...,115218.0,41.6,1.9,NaN,140.0,83137.0,41.9,2.4,NaN,313.0
3,01/04/2004,Aberdeenshire,S12000034,87373,42.5,NaN,1.0,NaN,NaN,571.0,...,115247.0,41.6,0.0,NaN,180.0,84241.0,42.5,1.3,NaN,391.0
4,01/05/2004,Aberdeenshire,S12000034,89493,43.5,NaN,2.4,NaN,NaN,502.0,...,117377.0,42.4,1.8,NaN,167.0,86466.0,43.6,2.6,NaN,335.0


In [254]:
data.describe()

,AveragePrice,Index,IndexSA,1m%Change,12m%Change,AveragePriceSA,SalesVolume,DetachedPrice,DetachedIndex,Detached1m%Change,...,NewPrice,NewIndex,New1m%Change,New12m%Change,NewSalesVolume,OldPrice,OldIndex,Old1m%Change,Old12m%Change,OldSalesVolume
count,1.490850e+05,149085.000000,5244.000000,148632.000000,144225.000000,5244.000000,144618.000000,1.422840e+05,142284.000000,141889.000000,...,1.416640e+05,141664.000000,141268.000000,136936.000000,138483.000000,1.418680e+05,141868.000000,141472.000000,137140.000000,141858.000000
mean,1.831012e+05,59.835763,62.132113,0.526866,6.100826,168752.656941,1252.600105,3.420380e+05,59.933395,0.461848,...,2.196692e+05,61.681981,0.497958,6.245756,135.956161,1.847985e+05,60.045649,0.472946,6.054274,1138.200045
std,1.311841e+05,26.619737,25.039764,1.934405,8.622357,94882.354669,7707.191260,3.432452e+05,25.206280,1.697813,...,1.342548e+05,26.401143,1.751680,7.974581,843.840990,1.310111e+05,25.507217,1.730925,8.408653,6965.933714
min,2.553000e+03,0.800000,13.900000,-30.300000,-35.800000,39008.000000,2.000000,4.011700e+04,9.400000,-19.600000,...,2.244300e+04,8.000000,-29.800000,-28.900000,1.000000,2.271600e+04,9.000000,-30.600000,-30.300000,2.000000
25%,1.005410e+05,38.200000,46.600000,-0.500000,1.100000,116135.000000,140.000000,1.785228e+05,41.200000,-0.500000,...,1.279912e+05,42.400000,-0.500000,1.700000,10.000000,1.021160e+05,40.900000,-0.500000,1.000000,126.000000
50%,1.547560e+05,61.000000,62.600000,0.500000,5.100000,153018.500000,215.000000,2.603840e+05,60.000000,0.400000,...,1.982200e+05,61.600000,0.500000,5.400000,21.000000,1.549365e+05,60.800000,0.500000,5.100000,193.000000
75%,2.301380e+05,80.300000,80.400000,1.400000,10.100000,209241.750000,376.000000,3.961765e+05,79.800000,1.400000,...,2.780390e+05,82.800000,1.500000,10.100000,46.000000,2.312262e+05,79.600000,1.400000,10.200000,343.000000
max,1.656986e+06,153.100000,110.000000,35.300000,98.400000,574627.000000,183609.000000,5.820174e+06,147.000000,34.800000,...,1.414204e+06,149.600000,34.900000,96.200000,21097.000000,1.665089e+06,152.700000,35.600000,98.600000,166098.000000


#### Review missing variables


In [256]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 149085 entries, 0 to 149084
Data columns (total 54 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   Date                    149085 non-null  str    
 1   RegionName              149085 non-null  str    
 2   AreaCode                149085 non-null  str    
 3   AveragePrice            149085 non-null  int64  
 4   Index                   149085 non-null  float64
 5   IndexSA                 5244 non-null    float64
 6   1m%Change               148632 non-null  float64
 7   12m%Change              144225 non-null  float64
 8   AveragePriceSA          5244 non-null    float64
 9   SalesVolume             144618 non-null  float64
 10  DetachedPrice           142284 non-null  float64
 11  DetachedIndex           142284 non-null  float64
 12  Detached1m%Change       141889 non-null  float64
 13  Detached12m%Change      137568 non-null  float64
 14  SemiDetachedPrice       142284 

We can see there are 53 columns, and a lot have missing values. We likely only the first 4-5 columns though

### Investigating area codes

In [259]:
unique_area_codes = data['AreaCode'].unique()
print(set([i[0] for i in unique_area_codes]))

{'N', 'K', 'S', 'W', 'E'}


In [260]:
print(set([i[0:3] for i in unique_area_codes]))

{'E11', 'K04', 'E12', 'N92', 'E13', 'E07', 'E10', 'E08', 'S12', 'E06', 'E09', 'S92', 'W06', 'N09', 'E92', 'K02', 'W92', 'K03'}


In [261]:
total_area_codes = len(unique_area_codes)
idx = 0
while idx <= total_area_codes:
    print(unique_area_codes[idx:idx+10])
    idx = idx+10
    

<StringArray>
['S12000034', 'E07000223', 'E07000032', 'S12000041', 'N09000001', 'N09000011',
 'S12000035', 'N09000002', 'E07000224', 'E07000170']
Length: 10, dtype: str
<StringArray>
['E07000105', 'E07000200', 'E09000002', 'E09000003', 'E08000038', 'E07000066',
 'E07000084', 'E07000171', 'E06000022', 'E06000055']
Length: 10, dtype: str
<StringArray>
['N09000003', 'E09000004', 'E08000025', 'E07000129', 'E06000008', 'E06000009',
 'W06000019', 'E07000033', 'E08000001', 'E07000136']
Length: 10, dtype: str
<StringArray>
['E06000058', 'E06000036', 'E08000032', 'E07000067', 'E07000143', 'E09000005',
 'E07000068', 'W06000013', 'E06000043', 'E07000144']
Length: 10, dtype: str
<StringArray>
['E09000006', 'E07000234', 'E07000095', 'E07000172', 'E06000060', 'E07000117',
 'E08000002', 'W06000018', 'E08000033', 'E07000008']
Length: 10, dtype: str
<StringArray>
['E10000003', 'E09000007', 'E07000192', 'E07000106', 'W06000015', 'W06000010',
 'E07000069', 'N09000004', 'E06000056', 'W06000008']
Length: 1

Area codes appear to be in format N-01-000-001, where the first letter is a country code, next two digits a regional code then last 3 a county / unitary / district authority or london borough.

There are country codes; E,K,N,S,W. These could be England, ?, Northern Ireland, Scotland and Wales.
The secondary codes for E are : 06 to 13 and 92.
The secondary codes for K are 02-04.
The secondary codes for N are 09,92.
The secondary codes for S are 12.
The secondary codes for W are 06,92.

In [263]:
### create a region column
data['Country'] = data['AreaCode'].str[0]

#### Create a new dataframe from key columns

For now we will use only the columns; date, region name, averagePrice and country. There are no missing values in these columns

In [265]:
data_s = data[['Date','RegionName','AreaCode','AveragePrice','Country']]
data_s.head()

,Date,RegionName,AreaCode,AveragePrice,Country
0,01/01/2004,Aberdeenshire,S12000034,84638,S
1,01/02/2004,Aberdeenshire,S12000034,84623,S
2,01/03/2004,Aberdeenshire,S12000034,86536,S
3,01/04/2004,Aberdeenshire,S12000034,87373,S
4,01/05/2004,Aberdeenshire,S12000034,89493,S


In [277]:
data_s.info()

<class 'pandas.DataFrame'>
RangeIndex: 149085 entries, 0 to 149084
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype
---  ------        --------------   -----
 0   Date          149085 non-null  str  
 1   RegionName    149085 non-null  str  
 2   AreaCode      149085 non-null  str  
 3   AveragePrice  149085 non-null  int64
 4   Country       149085 non-null  str  
dtypes: int64(1), str(4)
memory usage: 5.7 MB


### Investigating the area codes further

Now we have country codes, what region names are in each country?

In [171]:
group = data_s.groupby('Country')['RegionName'].unique()
print(group.values)

[<StringArray>
 [                    'Adur',             'Amber Valley',
                      'Arun',                 'Ashfield',
                   'Ashford',                  'Babergh',
      'Barking and Dagenham',                   'Barnet',
                  'Barnsley',                 'Basildon',
  ...
                 'Wokingham',            'Wolverhampton',
                 'Worcester',           'Worcestershire',
                  'Worthing',                 'Wychavon',
                      'Wyre',              'Wyre Forest',
                      'York', 'Yorkshire and The Humber']
 Length: 334, dtype: str
 <StringArray>
 ['England and Wales', 'Great Britain', 'United Kingdom']
 Length: 3, dtype: str
 <StringArray>
 [            'Antrim and Newtownabbey',                 'Ards and North Down',
  'Armagh City Banbridge and Craigavon',                             'Belfast',
             'Causeway Coast and Glens',             'Derry City and Strabane',
                  'Ferm

__We can see the region letter (i.e. the area code first letter), refers to England (E), Northern Ireland (N), 
Scotland (S) and Wales (W), not in fact North, South, East, West.__

__It may be more useful to split counties into regions.
The below csv includes a list of counties and their associated regions, from another government source.__

In [180]:
region_counties = pd.read_csv("../data/uk-counties-to-regions.csv")
region_counties.head()

,County,Region
0,Guernsey,Crown Dependencies
1,IOM,Crown Dependencies
2,Jersey,Crown Dependencies
3,North East Derbyshire,East Midlands
4,Amber Valley,East Midlands


In [218]:
### show list of regions from counties list

regions = region_counties['Region'].unique()
print(regions)

<StringArray>
[      'Crown Dependencies',            'East Midlands',
          'East of England',                   'London',
       'North East England',       'North West England',
         'Northern Ireland',                 'Scotland',
       'South East England',       'South West England',
                    'Wales',            'West Midlands',
 'Yorkshire and the Humber']
Length: 13, dtype: str


In [200]:
counties = region_counties['County'].tolist()
print("Counties",len(counties))

region_names = data_s['RegionName'].unique().tolist()
print("Unique region names",len(region_names))


Counties 531
Unique region names 405



__There are 405 region names in our original dataset. But the list of counties and unitary from a different source gives us 531 counties.__


#### List region names not in counties list

In [208]:
region_names_not_in_counties = set(region_names)-set(counties)
print(len(region_names_not_in_counties))
print(region_names_not_in_counties)

52
{'City of Plymouth', 'City of Aberdeen', 'Fermanagh and Omagh', 'Causeway Coast and Glens', 'Scotland', 'Na h-Eileanan Siar', 'Scottish Borders', 'East Suffolk', 'Inner London', 'West Midlands Region', 'West Suffolk', 'Mid and East Antrim', 'Derry City and Strabane', 'United Kingdom', 'Westmorland and Furness', 'Antrim and Newtownabbey', 'Newcastle upon Tyne', 'Bournemouth Christchurch and Poole', 'North East Derbyshire', 'Newry Mourne and Down', 'Mid Ulster', 'Great Britain', 'Rhondda Cynon Taf', 'City of Derby', 'Yorkshire and The Humber', 'East of England', 'North East', 'City of Glasgow', 'Folkestone and Hythe', 'West Northamptonshire', 'England and Wales', 'City of Kingston upon Hull', 'East Midlands', 'Outer London', 'Northern Ireland', 'Ards and North Down', 'City of Peterborough', 'Vale of Glamorgan', 'South West', 'Armagh City Banbridge and Craigavon', 'Lisburn and Castlereagh', 'North West', 'City of Dundee', 'City of Nottingham', "King's Lynn and West Norfolk ", 'Wales', 

Not as simple as we would of liked. There are some names in region_names that are not in our list of counties.
There is a map available at: https://geoportal.statistics.gov.uk/documents/1aa806eb35ee4334a87f5970c82e3ac0/explore

There is a list available at: https://assets.publishing.service.gov.uk/media/603901d7e90e070566dafa58/Regions_and_local_authorities_at_01-04-21.pdf

I manually use the above data to create the dictionary below.

In [225]:
region_names_to_regions = {'City of Plymouth':'South West England', 'City of Aberdeen':'Scotland', 'Fermanagh and Omagh':'Northern Ireland', 
                           'Causeway Coast and Glens':'Northern Ireland', 'Scotland':'Scotland', 'Na h-Eileanan Siar':'Scotland', 
                           'Scottish Borders':'Scotland', 'East Suffolk':'East of England', 'Inner London':'London', 
                         'West Midlands Region':'West Midlands', 'West Suffolk':'East of England', 'Mid and East Antrim':'Northern Ireland', 
                         'Derry City and Strabane':'Northern Ireland', 'United Kingdom':'none', 'Westmorland and Furness':'North West England', 
                         'Antrim and Newtownabbey':'Northern Ireland', 'Newcastle upon Tyne':'North East England', 
                           'Bournemouth Christchurch and Poole':'South West England', 
                         'North East Derbyshire':'East Midlands', 'Newry Mourne and Down':'Northern Ireland', 
                           'Mid Ulster':'Northern Ireland', 'Great Britain':'none', 
                         'Rhondda Cynon Taf':'Wales', 'City of Derby':'East Midlands', 
                           'Yorkshire and The Humber':'Yorkshire and the Humber', 'East of England':'East of England', 
                         'North East':'North East England', 'City of Glasgow':'Scotland', 'Folkestone and Hythe':'South East England', 
                           'West Northamptonshire':'East Midlands', 'England and Wales':'none', 
                         'City of Kingston upon Hull':'Yorkshire and the Humber', 'East Midlands':'East Midlands', 'Outer London':'London', 
                         'Northern Ireland':'Northern Ireland', 'Ards and North Down':'Northern Ireland', 
                         'City of Peterborough':'East of England', 'Vale of Glamorgan':'Wales', 'South West':'South West England', 
                         'Armagh City Banbridge and Craigavon':'Northern Ireland', 
                         'Lisburn and Castlereagh':'Northern Ireland', 'North West':'North West England', 
                         'City of Dundee':'Scotland', 'City of Nottingham':'East Midlands', 
                         "King's Lynn and West Norfolk ":'East of England' , 'Wales':'Wales', 
                         'Richmond upon Thames':'London', 'England':'none', 'South East':'South East England', 
                         'North Northamptonshire':'East Midlands', 'London':'London', 'Cumberland':'North West England'}




### Convert our dictionary of missing counties to a dataframe

In [324]:
missing_counties = pd.DataFrame.from_dict(region_names_to_regions,orient='index')
missing_counties = missing_counties.reset_index(drop=False)
missing_counties.columns = ['County','Region']
missing_counties.head()

,County,Region
0,City of Plymouth,South West England
1,City of Aberdeen,Scotland
2,Fermanagh and Omagh,Northern Ireland
3,Causeway Coast and Glens,Northern Ireland
4,Scotland,Scotland


In [335]:
# concatenate into one lookup dataframe

print(region_counties.shape)
print(missing_counties.shape)

all_counties = pd.concat([region_counties,missing_counties])

print(all_counties.shape)
print(all_counties.tail())

(531, 2)
(52, 2)
(583, 2)
                    County              Region
47                 England                none
48              South East  South East England
49  North Northamptonshire       East Midlands
50                  London              London
51              Cumberland  North West England


In [363]:

#### DUPLICATES EXISTS HERE

print(len(all_counties['County'].unique()))

554


__Now we have a mapping from counties to region in a separate dataframe. But the counties are (confusingly) called RegionName in our original dataset so we need to prepare for merge.__

### Merge the dataframes, data_s and all_counties to get a region column

In [343]:
# rename column in original dataset to county
data_s.rename(columns= {"RegionName":"County"},inplace=True)
data_s.head()

,Date,County,AreaCode,AveragePrice,Country
0,01/01/2004,Aberdeenshire,S12000034,84638,S
1,01/02/2004,Aberdeenshire,S12000034,84623,S
2,01/03/2004,Aberdeenshire,S12000034,86536,S
3,01/04/2004,Aberdeenshire,S12000034,87373,S
4,01/05/2004,Aberdeenshire,S12000034,89493,S


In [347]:
data_s = data_s.merge(all_counties, on='County',how='left')
data_s.head()


,Date,County,AreaCode,AveragePrice,Country,Region_x,Region_y
0,01/01/2004,Aberdeenshire,S12000034,84638,S,Scotland,Scotland
1,01/02/2004,Aberdeenshire,S12000034,84623,S,Scotland,Scotland
2,01/03/2004,Aberdeenshire,S12000034,86536,S,Scotland,Scotland
3,01/04/2004,Aberdeenshire,S12000034,87373,S,Scotland,Scotland
4,01/05/2004,Aberdeenshire,S12000034,89493,S,Scotland,Scotland


### RESOLVE THIS NEXT!!!

__have tried to merge on county but there appears to be some duplicate mappings for county-> region from all_counties. there is some duplicates in all_counties which i suspect produce different mappings for the same counties.__

In [357]:
bool_idx = data_s['Region_x'] != data_s['Region_y']
print(data_s[bool_idx].shape)

(2976, 7)


.
.
.
.
.
.
.
.
.
.


In [371]:
'''
east_of_england = ['Babergh', 'Basildon', 'Bedford', 'Braintree', 'Breckland', 'Brentwood', 'Broadland',
                'Broxbourne', 'Cambridge', 'Castle Point', 'Central Bedfordshire', 'Chelmsford',
                'Colchester', 'Dacorum', 'East Cambridgeshire', 'East Hertfordshire', 'East Suffolk',
                'Epping Forest', 'Fenland', 'Great Yarmouth', 'Harlow', 'Hertsmere', 'Huntingdonshire',
                'Ipswich', 'Kings Lynn and West Norfolk', 'Luton', 'Maldon', 'Mid Suffolk', 'North Hertfordshire', 
                   'North Norfolk', 'Norwich', 'Peterborough', 'Rochford', 'South Cambridgeshire', 'South Norfolk', 
                   'Southend-on-Sea', 'St. Albans', 'Stevenage', 'Tendring',
                'Three Rivers', 'Thurrock', 'Uttlesford', 'Watford', 'Welwyn Hatfield', 'West Suffolk']

# check diff
east_diff = set(east_of_england) - set(region_names)
print(east_diff)

east_midlands = ['Amber Valley', 'Ashfield', 'Bassetlaw', 'Blaby', 'Bolsover', 'Boston', 'Broxtowe',
                'Charnwood', 'Chesterfield', 'Derby', 'Derbyshire Dales', 'East Lindsey', 'Erewash',
                'Gedling', 'Harborough', 'High Peak', 'Hinckley and Bosworth', 'Leicester', 'Lincoln',
                'Mansfield', 'Melton', 'Newark and Sherwood', 'North East Derbyshire', 'North Kesteven', 'North West Leicestershire', 
                 'North Northamptonshire', 'Nottingham','Oadby and Wigston', 'Rushcliffe', 'Rutland', 'South Derbyshire', 
                 'South Holland', 'South Kesteven', 'West Lindsey', 'West Northamptonshire']

# check diff
east_m_diff = set(east_midlands) - set(region_names)
print(east_m_diff)
'''

"\neast_of_england = ['Babergh', 'Basildon', 'Bedford', 'Braintree', 'Breckland', 'Brentwood', 'Broadland',\n                'Broxbourne', 'Cambridge', 'Castle Point', 'Central Bedfordshire', 'Chelmsford',\n                'Colchester', 'Dacorum', 'East Cambridgeshire', 'East Hertfordshire', 'East Suffolk',\n                'Epping Forest', 'Fenland', 'Great Yarmouth', 'Harlow', 'Hertsmere', 'Huntingdonshire',\n                'Ipswich', 'Kings Lynn and West Norfolk', 'Luton', 'Maldon', 'Mid Suffolk', 'North Hertfordshire', \n                   'North Norfolk', 'Norwich', 'Peterborough', 'Rochford', 'South Cambridgeshire', 'South Norfolk', \n                   'Southend-on-Sea', 'St. Albans', 'Stevenage', 'Tendring',\n                'Three Rivers', 'Thurrock', 'Uttlesford', 'Watford', 'Welwyn Hatfield', 'West Suffolk']\n\n# check diff\neast_diff = set(east_of_england) - set(region_names)\nprint(east_diff)\n\neast_midlands = ['Amber Valley', 'Ashfield', 'Bassetlaw', 'Blaby', 'Bolsover'

### Use the government uk counties and unitary boundaries map
https://www.data.gov.uk/dataset/85228aec-fe0e-49bf-9455-df000d61e731/counties-and-unitary-authorities-may-2023-boundaries-uk-bgc

In [129]:
import geopandas as gpd
geodata = "../data/uk-counties.geojson"
map_df = gpd.read_file(geodata)
print(map_df.columns)

Index(['FID', 'CTYUA23CD', 'CTYUA23NM', 'CTYUA23NMW', 'BNG_E', 'BNG_N', 'LONG',
       'LAT', 'GlobalID', 'geometry'],
      dtype='str')
